### ISSUE TITLE
Add a maxspeed overlay

### ISSUE THREAD

**Issue by letypequividelespoubelles** [streetcomplete/StreetComplete#4282]
**Use case**
IMHO it would be a very usefull overlay, as maxspeed isn't a resurveyable quest, and not so easy to map and correct (no opendata for it in France, no easy editors to do it in situ). 

**Proposed Solution**
As UX, I might proposed that highway are in a specific color (red ?) if there are no maxspeed tag.
If a mawspeed is already tagged, put a sign (in a form of a traffic sign) to tell what's already tagged :
- speed limit value in a circle in case of maxspeed=* tag
- speed limit value in a square in case of a zone:mawspeed=* tag
- a city icon in case of a maxspeed=XX:urban tag
- a strikethrought city icon in case of a maxspeed=XX:rural

When clicking on a highway to add/modify the speed limit, the same UX as for the maxspeed quest would be perfect.

Thanks !

**Comment by westnordost** [comment]
Regarding the coloring, I've been doing some research and thinking about which maxspeed values could/should be grouped together. Though, the color suggestions are probably not color-blind-friendly.

 | % of maxspeed values | maxspeed value | suggested color |
| -- | -- | -- |
 1% | 130 | violet |
 1% | 120 | blue |
 1% | 110 | blue |
 3% | 100 | teal |
 3% | 90 | teal |
 6% | 80 | green |
 5% | 70 | green |
 8% | 60 | chartreuse |
27% | 50 | yellow |
 9% | 40 | yellow |
22% | 30 | amber |
 3% | 20 | orange |
 1% | 10 | orange |

**Comment by riQQ** [comment]
Here's the table with the colors visualized with example values:

| % of maxspeed values | maxspeed value | suggested color | color code |
| -------- | ---------- | ------------ | ---- |
| 1% 	| 130 | violet | `#ee82ee` |
| 1%  | 120 | blue | `#0000ff` |
| 1%  | 110 | blue | `#0000ff` |
| 3%  | 100 | teal | `#008080` |
| 3%  | 90  | teal | `#008080` |
| 6%  | 80  | green | `#008000` |
| 5%  | 70  | green | `#008000` |
| 8%  | 60  | chartreuse | `#7fff00` |
| 27% | 50  | yellow | `#ffff00` |
| 9%  | 40  | yellow | `#ffff00` |
| 22% | 30  | amber | `#ffbf00` |
| 3%  | 20  | orange | `#ffa500` |
| 1%  | 10  | orange | `#ffa500` |

<details>
  <summary>Click for markdown</summary>

```
| % of maxspeed values | maxspeed value | suggested color | color code |
| -------- | ---------- | ------------ | ---- |
| 1% 	| 130 | violet | `#ee82ee` |
| 1%  | 120 | blue | `#0000ff` |
| 1%  | 110 | blue | `#0000ff` |
| 3%  | 100 | teal | `#008080` |
| 3%  | 90  | teal | `#008080` |
| 6%  | 80  | green | `#008000` |
| 5%  | 70  | green | `#008000` |
| 8%  | 60  | chartreuse | `#7fff00` |
| 27% | 50  | yellow | `#ffff00` |
| 9%  | 40  | yellow | `#ffff00` |
| 22% | 30  | amber | `#ffbf00` |
| 3%  | 20  | orange | `#ffa500` |
| 1%  | 10  | orange | `#ffa500` |
```

</details>

**Comment by westnordost** [comment]
Though, the overlay needs to handle situations where both the maxspeed and the implict maxspeed road type is tagged, too. And maybe other less common situations.

**Comment by letypequividelespoubelles** [comment]
How is it possible ? You can't have for exemple maxspeed=FR:urban and maxspeed=50 on the same way, if yes it's a mistake ! Do you have any possible right tagging scheme where this might happen ?

**Comment by westnordost** [comment]
But you can have `maxspeed:type=FR:urban` and `maxspeed=50´ on the same way.

**Comment by matkoniecz** [comment]
> Though, the overlay needs to handle situations where both the maxspeed and the implict maxspeed road type is tagged, too. And maybe other less common situations.

In case of matching ones it can be ignored, and in case of mismatches I think it would go to "I give up" class.

But that would require maintaining list of expected maxspeed values...

**Comment by westnordost** [comment]
The colors above are obsolete because there'd need to be a different color palette anyway (that is color-blind friendly)

**Comment by joshinils** [comment]
This should also tag the source, e.g. a zone vs a plain sign.
Examples include: `source:maxspeed=DE:zone30`, `maxspeed=30` + `maxspeed:type=sign`, `maxspeed:type=DE:zone30`, `source:maxspeed=DE:bicycle_road`, `zone:maxspeed=DE:30 `, `traffic_sign=DE:274.1`, `traffic_sign=DE:274-30`, `traffic_sign=DE:244.1`, `traffic_sign=DE:244.3`

and which of these are mutually exclusive...

what about tagging maxspeed as `walk`? for example where `traffic_sign=DE:239,1022-10` implies cycling at walking pace (i.e. `maxspeed:bicycle=walk`) or `highway=living_street` in germany too.

**Comment by matkoniecz** [comment]
The highest complexity would be likely for cases where speed limits depends on way direction. 

Or https://taginfo.openstreetmap.org/keys/maxspeed%3Aforward#overview https://taginfo.openstreetmap.org/keys/maxspeed%3Abackward#overview  is edge case that can be ignored?

2025 EDIT: HAHAHAHAAHAHA, past me thing it will be the worst part was so cute and naive. That is only start of the nightmare

**Comment by tszym** [comment]
Even considering that as edge case to ignore for a first implementation, the overlay would bring a lot of help to mappers already.
Eventually it could be nice to handle direction dependent speed limits, maybe with an UI similar to the one for the sidewalks overlay (maybe hidden behind a "other answers" button).

I believe the implementation of that overlay should not be blocked by discussions over trying to nail the perfect feature covering all cases right from the beginning, as from a user perspective, it feels like that overlay would help a lot even with basic features, while more improvements are developed later.
Max speeds are a complex issue, especially since it is country dependent, so it could even be first developed behind a regional feature flag, like some quests that are never displayed for some countries.

**Comment by arrival-spring** [comment]
Colours seem to be the tricky thing here, as there are lots of possible maxspeed values and, at the moment, only 9 colours in the overlay palette.

The JOSM [maxspeed style](https://josm.openstreetmap.de/wiki/Styles/Maxspeed) has this:
![Chart of maxspeed values and their colouring](https://josm.openstreetmap.de/raw-attachment/wiki/Styles/Maxspeed/maxspeed_legend.png)

[Colourblind comparison of these colours](https://davidmathlogic.com/colorblind/#%23febede-%23ff5656-%2300da8b-%23e3cc82-%234d4dff-%234d4d00-%23ff529a-%2307f807-%239f9fff-%23ee8b00-%23f90808-%2307f8f8-%23f807f8-%237807f8-%23b69b20-%23c0f100-%23438100-%23febede)

**Comment by RubenKelevra** [comment]
@westnordost wrote:
> But you can have `maxspeed:type=FR:urban` and `maxspeed=50` on the same way.

That's why I prefer to not spread the same information to two tags.

The OSM wiki got a machine readable list of the implicit `maxspeed=*` values. So StreetComplete could compare the explicit value of the maxspeed tag to what's tagged in the depricated `source:maxspeed` / `maxspeed:type` Tags.

If there's a missmatch the road can be rendered with red as an error (without the overlay SC could show a quest for maxspeed here).

Then both options can be shown to the user to decide which one is correct and the tags can be cleaned up by either tagging `maxspeed=50` (and `maxspeed:type=sign` if you think that's necessary) or `maxspeed=FR:urban`.

With a third button the user could respond that both is wrong and default back to the regular maxspeed questionare.

StreetComplete could do the same for resurveying, so the tags are slowly fixed if surveyed by users.

**Comment by RubenKelevra** [comment]
@arrival-spring wrote:
> Colours seem to be the tricky thing here, as there are lots of possible maxspeed values and, at the moment, only 9 colours in the overlay palette.
> 
> The JOSM [maxspeed style](https://josm.openstreetmap.de/wiki/Styles/Maxspeed) has this: ![Chart of maxspeed values and their colouring](https://camo.githubusercontent.com/48660e9291687eae9d4b93a50ada144b23670ef15860509450ca4ded3761f1d4/68747470733a2f2f6a6f736d2e6f70656e7374726565746d61702e64652f7261772d6174746163686d656e742f77696b692f5374796c65732f4d617873706565642f6d617873706565645f6c6567656e642e706e67)
> 
> [Colourblind comparison of these colours](https://davidmathlogic.com/colorblind/#%23febede-%23ff5656-%2300da8b-%23e3cc82-%234d4dff-%234d4d00-%23ff529a-%2307f807-%239f9fff-%23ee8b00-%23f90808-%2307f8f8-%23f807f8-%237807f8-%23b69b20-%23c0f100-%23438100-%23febede)

I think the color scheme of JOSM is pretty suboptimal, as it includes different shades of red which is currently used to show missing information.

Maxspeed zones are also missing from that list... so there are additional visual cues or colors necessary for this...

I got a different idea:

How about using solid green for `xx:urban`, solid amber for `xx:rural` and solid orange for `xx:motorway`. 

For explicit numbers the path gets rendered as zebra with black.

So e.g. for Germany:
- `DE:urban` = green
- `50` = zebra green
- `DE:rural` = amber
- `100` = zebra amber

For numbers different than these the colors are mixed, depending on the distance - so `70` gets a 20:30 mix of green and amber.

To have colors to mix numbers below `xx:urban` is defined for `0` and violet is defined for `150` (and `none`) to allow for numbers greater than `xx:motorway`.

If there's a zone, the secondary color for the zebra could be changed to white.

This way we don't have to worry about mph values, as the color scheme just adopts to the local defaults and mixes from there.

It also gives some visual feedback if a speedlimit has increased or decreased by a 'severity' type color mixing - which works better visually IMHO than random colors schemes used in JOSM.

Btw: Color blindness shouldn't really be a consideration here, as Android supports color correction for the different types:

![Screenshot_2023-05-14-09-57-29-146-edit_com android settings](https://github.com/streetcomplete/StreetComplete/assets/614929/1ccbae47-38e8-48ce-8ac8-57370b18cb6f)

**Comment by joshinils** [comment]
> depricated `source:maxspeed` / `maxspeed:type` Tags.

Since when? The wiki does not mention them as deprecated but shows **Status: in use**.

**Comment by RubenKelevra** [comment]
> > depricated `source:maxspeed` / `maxspeed:type` Tags.
> 
> Since when? The wiki does not mention them as deprecated but shows **Status: in use**.

I was referring to the fact that you don't need an additional tag recording the "source of the maxspeed" if you tag the implicit value in the `maxspeed`-Tag directly.

So I get why you may want to add `maxspeed:type=sign` if you tag an explicit value, but I don't see a reason to add a `source:maxspeed=xx:urban` if you tag `maxspeed=xx:urban`.

**Comment by arrival-spring** [comment]
I have started work on this, and despite my previous comment, colours are probably not going to be the most difficult bit.

My work so far has been on a parser and tagger for speed limits (see here https://github.com/streetcomplete/StreetComplete/issues/1998#issuecomment-676787192 for some idea of the complications with that)

Next I've been thinking about the UI and would welcome input. If no speed limit is set we want to encourage people to say either that there is a sign (and its value) or that there is no sign and that the road is urban/rural/whatever. However, some people like to tag, say, both `maxspeed:type=FR:urban` and `maxspeed=50` so these would both need to be presented to the user and allow them to edit or remove them. And then there's the option of switching to `highway=living_street` as well.

And what to do about `maxspseed:advisory`? The maxspeed quest treats it as mutually exclusive with `maxspeed` or `maxspeed:type`, however about half of the time it's tagged in the wild it's combined with `maxspeed`, so should that be presented as well? As in "There's also an advisory speed limit", but then that's potentially three things for the user to be editing at once: type, limit and advisory limit.

I'm aware of [default speed limits](https://wiki.openstreetmap.org/wiki/Default_speed_limits) and [osm-legal-default-speeds](https://github.com/westnordost/osm-legal-default-speeds) and have thought about using it, but am not quite sure. Any suggestion of adding explicit speed limits when the user did not select it is out of the question, but maybe using it to attempt to show when `maxspeed` and `maxspeed:type` are mismatching could be ok, but would at least be quite complex.

When I mentioned the JOSM style I was by no means saying that it was the best option, just noting it as something that exists.

> How about using solid green for `xx:urban`, solid amber for `xx:rural` and solid orange for `xx:motorway`.

And there's also `xx:trunk` in some countries at least, as well as the GB tagging.

There is also the option on overlays of displaying text written along the way, but this would not be so visually striking for spotting errors.

**Comment by joshinils** [comment]
> However, some people like to tag, say, both `maxspeed:type=FR:urban` and `maxspeed=50`

I don't remember where (maybe at FOSSGIS), but I heard it is more important to have the explicit number in `maxspeed` over anything else, since some software only cares about that tag.
Anything else is nice to have of course, but I think there should never be no bare maxspeed tag.

**Comment by arrival-spring** [comment]
In SC people can only survey what's there, and if there's no sign then they can't tag maxspeed.

Due to lack of maxspeed tag in many places, software that solely relies on that needs to catch up, but I think all routing software makes assumptions, as you basically have to.

maxspeed:type is more information and is definitely useful

**Comment by NathanARF** [comment]
I would like to see an overlay for this have both `maxspeed` and `maxspeed:type` available to edit, particularly in the UK where there are a lot of speed limits mapped, but either `maxspeed:type` (useful for 60 mph and 30 mph speed limits) or `maxspeed` are not tagged. A lot of roads around me are lowering to 20 mph, which involves the changing of both tags, so the more we can alter and visualise on the go, the better in my opinion.

**Comment by RubenKelevra** [comment]
@NathanARF I disagree. StreetComplete has never shown any tags, it has always abstracted the technicalities and shown in a simple easy to approach way. If you want to see and edit the tags directly, just open the map in Vespucci?

**Comment by NathanARF** [comment]
I do not mean to edit tag values, as SC is meant to be for novice users, right? Just that the ability to make sure both `maxspeed` and `maxspeed:type` can be added or updated on road segments by users and visualised appropriately. By selecting a road segment with `maxspeed=30 mph` and `maxspeed:type=sign`, this could be shown with a sign graphic with the speed limit tagged on the road.

**Comment by RubenKelevra** [comment]
@arrival-spring wrote:
> In SC people can only survey what's there, and if there's no sign then they can't tag maxspeed.

This conclusion goes against common practice. The absence of a sign is also an information which can be mapped. Similar to the abscene of a "don't park here" sign where you're adding "edge of the road" parking via the "Street parking" overlay, if you're inside city limits (at least that's the law here in Germany).

So I don't think that's a plausible scenario that a user, who's local most of the time, cannot answer what default limit applies if he/she walks along a road without a posted speed limit on it.

**Comment by RubenKelevra** [comment]
> I do not mean to edit tag values, as SC is meant to be for novice users, right? Just that the ability to make sure both `maxspeed` and `maxspeed:type` can be added or updated on road segments by users and visualised appropriately. By selecting a road segment with `maxspeed=30 mph` and `maxspeed:type=sign`, this could be shown with a sign graphic with the speed limit tagged on the road.

I mean I think that's what I suggested before, isn't it? I just don't see why you want two layers make more sense here than one?!

SC should read all three tags which are currently commonly used and if they don't agree show this as an error. Otherwise they info should be presented to the users in a single layer.

If there are changed made, it makes sense to tag the speed into `maxspeed=*` as explicit or implicit value and if explicit add the `maxspeed:type=sign` Tag in addition.

**Comment by joshinils** [comment]
How much does SC want to rely on the person using it to know the local laws?

If I'm on holiday, I would have to disable this quest, or at least a part of it, since I may not know where or why there is which default speed limit.
In that sense, SC should only ask for obvious things like explicit signs or a lack thereof.
I may be able to answer that there is a city-limit sign, but not know that `traffic_sign=DE:310` implies `maxspeed=50` for the direction I am facing.

**Comment by RubenKelevra** [comment]
@NathanARF wrote:
> Perhaps asking a question if a certain sign is seen would be required then? The 'urban area' signs across Europe vary, and some do not change the speed limit.

What do you mean?

SC does exactly this for as long as I remember in the maxspeed quest. This ticket is about an overlay for maxspeed ...

![Screenshot_2023-05-15-20-04-12-942-edit_de.westnordost.streetcomplete.jpg](https://github.com/streetcomplete/StreetComplete/assets/614929/af4e5f4d-d8db-4ce9-8a40-58b4845d549a)

**Comment by joshinils** [comment]
Then the overlay has to show both _if_ a sign has been seen previously (or _not_) and at the same time display a potential number which has been seen on said sign.

Overall, It should say if one can deduce the actual maxspeed from the tags present or not, or if they are mismatched and in conflict.

**Comment by RubenKelevra** [comment]
@joshinils wrote:

> How much does SC want to rely on the person using it to know the local laws?\n\nIf I'm on holiday, I would have to disable this quest, or at least a part of it, since I may not know where or why there is which default speed limit.\nIn that sense, SC should only ask for obvious things like explicit signs or a lack thereof.\nI may be able to answer that there is a city-limit sign, but not know that traffic_sign=DE:310 implies maxspeed=50 for the direction the sign is facing.

SC always relies on local law knowledge. If you're not certain how something is restricted/working/designed/meant to you should not answer it.

I wouldn't go on holiday and start mapping parking situations or speed limits in a foreign country, because I know I have no idea how these work.

So yes, you would deactivate the quest. But how is that an argument against showing an overlay for speed data in SC?

**Comment by NathanARF** [comment]
> What do you mean?
> 
> SC does exactly this for as long as I remember in the maxspeed quest. This ticket is about an overlay for maxspeed ...

I was thinking more about asking whether an urban area sign is seen with regards to confusing limits in foreign countries, but you're right users should not attempt to solve quests they don't know much about

**Comment by westnordost** [comment]
@arrival-spring well, all these edge cases have been the reason why I shied away from implementing this so far. Not only is it an incredibly complex topic because there are so many possible ways to tag it, the whole `source:maxspeed` topic doesn't exactly make it easier. I recently made a list of things to take into account, I'll just put it here:

- need to support everything that is valid and may not overwrite such valid data, only correct or complement it: forward/backward, for different times, varying speed limits (through digital display), advisory speed limit, walk speed limit, etc. 

- should probably not replace `zone:maxspeed` (`source:maxspeed`, etc.) with `maxspeed:type` when re-tagging as long as none of these keys is deprecated but in that case just update the current tag. Problem: For some of these keys, `sign` is actually not a valid value. Problem 2: For some of these keys, the values for slow speed zones differ (`DE:30` vs `DE:zone30`)

- need to decide on a color palette for the speeds - which number ranges map to which color?

- instead of specifying speed limit or urban/rural, need to offer choices for streets with fixed speed limits based on another sign (living streets, bicycle roads?, pedestrian zone?, school zone?, silver zone?, shared zone?, ... more? - see default speed limits, depending on the country, there may be many of these). Of course, need (simplified) pictures of such signs too

- not all countries use urban/rural/... (or whatever) distinction. There is a huge range of other distinctions which is pretty much impossible to properly support. Reason: For many countries, there is not even a `source:maxspeed` value defined for certain road types. E.g. some US states have different speed limits for residential/business/other but there is not even any value defined for this in the wiki. Hence, not really sure how to solve this.

- As per previous point, `maxspeed:signed=yes/no` (or similar) would solve it all in one go and make it also easier for the surveyor. Maybe propose it? Unknown if there'd be support for this in the community. Does it need to be proposed? `*:signed=yes/no` keys are already well established - albeit mostly by StreetComplete itself.

- should probably support tagging vehicle specific speed limits (200k+ taggings)

- should probably support time-limited speed limits (60k+ taggings)

- should probably support speed limits that differ per direction (100k+ taggings)

**Comment by westnordost** [comment]
As for the UI, I imagine it like this:

1. Same UI as for lit overlay: Current state is displayed in tappable box. When tapped, a range of possibilities is displayed

2. Possibilities should include: living street, bicycle road, pedestrian street, ... etc. , digital display, sign, advisory sign, no sign, ... 

3. Certain selections, such as "sign" and "advisory sign" require input within the selected item. (I'd say, after the item has been selected, not in some dialog box)

4. If there is a fixed speed limit specified but the speed limit is supposed to be implicit (=`source:maxspeed` is something else than `sign`), the item that is shown should be "no sign", with no possibility for the user to input or change the speed limit. To keep the UI simple. If the fixed speed limit does not match with what the default-speed-limits would suggest, it should be treated as invalid / not set.

**Comment by arrival-spring** [comment]
It certainly is complex, and seems to get more so the more I think about all of the possible combinations and edge cases. I'm making progress, but it will be a while.

Thank you for that list, I had thought of some of that, and had not been sure whether it was worth implement some of it. I'll work through it and see how I go.

I've had some discussion on OSM Slack about USA tagging and will probably ask some more questions in various places too.

Not allowing to edit the speed limit would be at least a bit controversial I should think, but I do understand the reasoning.

**Comment by westnordost** [comment]
An additional note: Different speed limits per lane. I've noticed during my vacation in Spain in a number of different cities that on urban roads that have more than one lane in a direction, oftentimes one of these lanes is limited to 30 (and has bicycle "sharrow" markings) while the other has the normal 50 speed limit.

![20230524_202113](https://github.com/streetcomplete/StreetComplete/assets/4661658/78ef4cdc-6d90-4147-9209-d7ac28516a05)

(Hence the "recuerde", because it basically is the rule in the whole city.)

If not a direct support to let the user specify it, at least it must be handled somehow, e.g. if `maxspeed:lanes=*` is specified but `maxspeed=*` isn't, or the user specifies a `maxspeeed` that is not the `max` of the speed limits specified in `maxspeed:lanes`.

**Comment by joshinils** [comment]
> Isn't the correct answer for SC to leave a note?

No, if there is a tagging scheme available, use it.
In this case, there is: https://wiki.openstreetmap.org/wiki/Key:maxspeed#Lanes
We (as mappers by hobby) are not responsible for how the data is used and by whom.

Otherwise:
I don't like leaving notes If I don't know if something can even be tagged at all in the first place.
i.e. _Any answer the user can give must have an equivalent tagging._

**Comment by westnordost** [comment]
Both of your comments are off topic here.
 
Leaving a note in the maxspeed **quest** is fine for an uncommon situation (maxspeed hours, maxspeed different per side, maxspeed different per lane, etc etc.). After all, the quest is only asked *once* when no `maxspeed` data is yet tagged. (Also, just specifying the max speed, i.e. 50 in the case above, would not be wrong either.)

For the overlay it is a different situation, because it has to deal with data that is already there *somehow* plus offers the user to edit it. So, it has to make sure that it does not overwrite correct data with incorrect data or creates conflicting data (e.g. `maxspeed:lanes=50|30` and `maxspeed=30` is added and otherwise nothing is changed.)

**Comment by arrival-spring** [comment]
I've found a problem with using `osm-legal-default-speeds` to determine if the `maxspeed` is 'correct'.

e.g. in France there are rural roads and rural dual carriageways as different road types.
Problem:
* A road that is a rural dual carriageway could be correctly tagged as `maxspeed:type=FR:rural oneway=yes maxspeed=110`
* A rural single carriageway that is oneway, e.g. because it is just before a roundabout could be correctly tagged as `maxspeed:type=FR:rural oneway=yes maxspeed=80`

But whether `maxspeed` is included or not, this always exact matches as rural dual carriageway, so the latter would appear as a 'wrong' speed limit.
* [maxspeed:type=FR:rural oneway=yes]( https://westnordost.github.io/osm-legal-default-speeds/#tags=maxspeed%3Atype%3DFR%3Arural%0Aoneway%3Dyes&cc=FR)
* [maxspeed:type=FR:rural oneway=yes maxspeed=110](https://westnordost.github.io/osm-legal-default-speeds/#tags=maxspeed%3Atype%3DFR%3Arural%0Aoneway%3Dyes%0Amaxspeed%3D110&cc=FR)
* [maxspeed:type=FR:rural oneway=yes maxspeed=80](https://westnordost.github.io/osm-legal-default-speeds/#tags=maxspeed%3Atype%3DFR%3Arural%0Aoneway%3Dyes%0Amaxspeed%3D80&cc=FR)

`oneway=yes` fuzzy matches dual carriageway
`rural=yes + oneway=yes` exact matches rural dual carriageway though

How is this best solved? I would really want a list of all the possible speeds it could be, based on the fuzzy matching, but that's not quite the aim of `osm-legal-default-speeds`

**Comment by westnordost** [comment]
Yes, the analysis is correct.

Hm well, what exactly to do you want to use the default speed limits for?

**Comment by westnordost** [comment]
If the answer is: To see if the `maxspeed` matches with what would be the default speed limit in case an indicator is present that the `maxspeed` does not come from a sign: That's a bit difficult, because

- there could always be mistakes in that data, or the data could be outdated. We probably don't want to mark everything red because the data hasn't been updated
- I guess it would only be OK to use exact matches, not fuzzy matches to determine if the explicit and implicit speed limit don't match. Plus, it has to made sure that exact matches are really always exact matches, not 99% sure guesses. Until very recently, the `oneway=yes` thing was not even classified as a fuzzy match, it was classified as an exact match

**Comment by arrival-spring** [comment]
Indeed, it's as you suggested here:

> 4. If there is a fixed speed limit specified but the speed limit is supposed to be implicit (=`source:maxspeed` is something else than `sign`), the item that is shown should be "no sign", with no possibility for the user to input or change the speed limit. To keep the UI simple. If the fixed speed limit does not match with what the default-speed-limits would suggest, it should be treated as invalid / not set.

I've got to the point where I started thinking about that and came to similar conclusions as you.

To continue with my example above, both of those would be fine, but a road with `maxspeed=70` would be an error and so it would be good if that was highlighted somehow.

The alternative would be to display both the type "urban/rural..." and the speed value, although if I was visiting some other country I would have no idea if they were not matching.

And if we use default speeds but then some law changes then should all earlier versions be banned from editing to save from trying to 'fix' data which is actually because of outdated matching data held in that version of SC?

**Comment by westnordost** [comment]
I read this comment a week ago but couldn't come up with a better solution than showing both at least in the case where they do not match.

> And if we use default speeds but then some law changes then should all earlier versions be banned from editing to save from trying to 'fix' data which is actually because of outdated matching data held in that version of SC?

Earlier versions cannot be banned from editing, only from uploading. This mechanism wouldn't fit here.

**Comment by arrival-spring** [comment]
> > And if we use default speeds but then some law changes then should all earlier versions be banned from editing to save from trying to 'fix' data which is actually because of outdated matching data held in that version of SC?
> 
> Earlier versions cannot be banned from editing, only from uploading. This mechanism wouldn't fit here.

Hmm, I didn't make myself clear there. e.g. at the moment `GB:nsl_single` means `60 mph`, but if the law changed next year and it then meant `50 mph` then (earlier versions of) SC (before updates to `default-speed-limits`) would show `GB:nsl_single` + `50 mph` as an error or incomplete or something if we were in any way strict with using `default-speed-limits`. So we need to avoid that situation.

> I read this comment a week ago but couldn't come up with a better solution than showing both at least in the case where they do not match.

So maybe we need to **always** show both (as we cannot match type and speed with 100% confidence and forward/backward compatibility)?

i.e. For places with no data, user can select either sign or no sign
For places with only one of type/maxspeed tagged we just show that
For places with type **and** maxspeed tagged we show both (and allow both to be edited? But if user changes type to other unsigned type then only show that and remove existing speed).

**Comment by westnordost** [comment]
Maybe, yeah... do you have a clear idea how the UI should look like already?

**Comment by arrival-spring** [comment]
A reasonable idea, I haven't got to the stage of implementing it yet though.

It risks getting complicated when taking into account vehicles and conditions, and then that in each direction also.

* Main type/speed at the top
* Row/s for any conditional with smaller speed sign and drop down to select condition type (and then add details)
- * Conditionals e.g. depends on time of day, depends on time of year, depends on weather
* Row/s for vehicles, denoted by icons, then similarly any conditions to the side

If directions are specified then the streetside view would maybe just show the speed/type (could have a plus or something if there's more tagged) then tapping on a side takes you to a similar form to the above.

**Comment by westnordost** [comment]
Here are some mockups of how I think the basic (and more advanced) UI could look like, without considerations on how to display urban/rural + maxspeed sign at the same time:

![maxspeed-form-basic](https://github.com/streetcomplete/StreetComplete/assets/4661658/50053ab3-d215-43ed-9d0e-7c4f0ce22377)

![max-complexity](https://github.com/streetcomplete/StreetComplete/assets/4661658/975fb767-65c9-4638-a563-4134020310e3)

**Comment by matkoniecz** [comment]
@arrival-spring 

Is https://github.com/arrival-spring/StreetComplete/tree/maxspeed-overlay having latest code that you did? WHat you recommend using it as a base for someone who would want to finish it?

Are there somewhere TODOs of what remains unfixed (or maybe TODOs are in code?)?

**Comment by arrival-spring** [comment]
> Is https://github.com/arrival-spring/StreetComplete/tree/maxspeed-overlay having latest code that you did?

I have just pushed a couple of unpushed commits, so it is now up to date with my latest code.

> WHat you recommend using it as a base for someone who would want to finish it?

It was so much work analysing all of the tagging that yes, I think I would.

I had not got to any UI work, but was in fact still in parsing and tagging of speed limits, however that was largely complete.

The point where I got stuck was with supporting time conditional speed limits, as I could not fully understand the opening hours parser and how to add to it. Specifically, it needs to support at least "SH off" and also month only limits, "Jun-Dec" etc., if it is to support the majority of time conditional speed limits.

It's possible that the data structure I went for is too complicated, nested and confusing. Changing it would be possible, but somewhat involved.

The problem is that speed limits (in OSM) are so complicated. It is possible to have/tag a speed limit, which could be different per vehicle, depend on weight, weather or time, which could be different per side of the road and even per lane. There could also be an advisory limit. And then we want to know what type of speed limit it is (`maxspeed:type` etc.), again that could be per direction, but probably not applicable to the rest of it. I did deal with almost all of that complexity, but putting it into an accessible data structure is not trivial, and the one I have might not be.

Also, I have seen that some of the way tests are done has been changed on master since I did this work, so will need changing here as well.

> Are there somewhere TODOs of what remains unfixed (or maybe TODOs are in code?)?

I think there are a couple of TODOs in code, but main things are as above, and then actually making the overlay and designing the UI.

One TODO which I think is in the code is to remove the checks and tests for `maxspeed:seasonal:winter` etc. as I have completed a (discussed) mass edit to retag such things to use `maxspeed:conditional` and so support is no longer needed.

**Comment by matkoniecz** [comment]
I guess the question at which point StreetComplete should give up and announce that it is too complex to edit with it.

For example conditional per lane speed limits may be a point where detecting and giving up would be good enough.

**Comment by tiptoptom** [comment]
By the way, you can quite easily configure a maxspeed overlay in SCEE.

![002](https://github.com/user-attachments/assets/04f4a5b3-5470-4685-aa9c-421c26e43310) 
![001](https://github.com/user-attachments/assets/0188619d-1321-4687-88bf-95e5abdf1bcf)

**Comment by RubenKelevra** [comment]
@tiptoptom yeah, sure this works for one tag. But there are several competing tagging standards and also there's maxspeed:forward/maxspeed:backward as well as ":conditional", which all would appear as red.

**Comment by mnalis** [comment]
> But there are several competing tagging standards and also there's maxspeed:forward/maxspeed:backward as well as ":conditional",

SCEE custom overlays supports full `elementFilter` syntax (e.g. `maxspeed or maxspeed:type or maxspeed:forward or maxspeed:backward or maxspeed:conditional or maxspeed:forward:conditional or maxspeed:backward:conditional` or whathaveyou) which solves that particular complaint; but if people are interested in how to use it, that should really be discussed in [SCEE repo](https://github.com/Helium314/SCEE/) instead.

Having a special dedicated overlay like suggested here would bring several advantages (for one, nice UI for answers like in existing MaxSpeed Quest; instead of having to know tags by heart to add/edit them in text format, as SCEE Custom Overlay needs).

**Comment by RubenKelevra** [comment]
Ah, thanks, didn't know that!

**Comment by matkoniecz** [comment]
replaced by https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-3133408120

posted https://community.openstreetmap.org/t/proposing-an-new-speed-limit-implied-yes-tag-for-roads-with-default-speed-limits-where-for-some-reason-mapper-was-unable-to-be-more-specific/132700 - looks that it got rejected

I started reviewing discussion that happened so far, to list what is required to do this behemoth.
I processed large part, but not all, as I realised that without some way to mark "no explicit speed limit sign here, not sure which default rule applies" this seems not doable.

-------------------------

I drafted post for OSM forums
<details>
<summary>Text of post draft, feedback welcome, I plan to post it tomorrow</summary>

title: Proposing an new speed_limit_implied=yes tag for roads with default speed limits where for some reason mapper was unable to be more specific

Sorry for a long post, but topic is tricky. If anyone manages to read entire one I would appreciate it.

Surveying speed limits is useful but horrifyingly complex.
Even seemingly clear cases, where numeric speed limit are signed, has problem with correctly marking where their applicability ends.

[Aktron / Wikimedia Commons - CC-BY-SA-4.0](https://commons.wikimedia.org/wiki/File:Raszowa,_omezen%C3%AD_rychlosti.jpg)

In some countries it may apply under next crossing - with extra fun what counts as a crossing for this legal purpose.
In some "until further notice" which can happen in multiple ways.
Many have also zone signs, which are cancelled by end zone sign and regular expiration for other signs.

You can get speed limits differing depending on lane, depending on vehicle type, different in forward and backward direction on the same way.
And combinations of that, and unlimited complexity of maxspeed:conditional (paging @minh_nguyen for some horrific but technically still applying speed limits).

Large speed limit zones may make surveying quite complex. Sometimes they can be quite confusing.

But overall signed speed limit or signed allowed unlimited speed is fairly doable.

--------------------------------

But you also get cases which are worse. Where you have no signed speed limit, but it is left implied and falls back to default speed limit defined in legislation.

On borders of Poland following "helpful" default speed limit welcomes travelers:

https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/Znak_D-39_(wersja_2021).svg/960px-Znak_D-39_(wersja_2021).svg.png 
https://commons.wikimedia.org/wiki/File:Znak_D-39_(wersja_2021).svg

[it used to be even worse](https://commons.wikimedia.org/wiki/File:Znak_D-39._Ograniczenia_pr%C4%99dko%C5%9Bci_w_Polsce_od_2011.svg)

It would be nice if mapper could answer "speed limit is implied here" if not sure which specific rule applies, rather than tag something wrong or incomplete or be unable to tag anything.

Obviously, ATYL applies but in my case I want to add support for such tagging in editor (StreetComplete) and I do not want to do this if OSM community disagrees.

And when making an editor there are some cases where asking user about exact speed limit is likely to result in nonsense answer, even after putting massive effort into interface guiding mapper.

For example in United States where speed limit is almost always signed with numeric speed limit signs. Signs such as "maximum speed 75" are basically always present. As I understand, speed limit falling back to default speed limits set by legislation are very rare. The problem is that it is not merely built-up / outside settlements distinction like in much of Europe. It is a bizarre mix of piles of rules, defined in various government levels. Due to their immense complexity and rarity how often they apply, noone bothers to learn them. 
Just collecting all of them and monitoring law for changes works would probably require full time work of multiple people.
And that is not the worst part - many of them are poorly defined, and there is no way to ask mapper in way that anyone would answer it correctly.

For example, in Colorado: what is in legal terms threshold between "rural winding mountain road" and "rural open mountain road"?

And “Residence district” in Iowa that has own default speed limits on roads within it has specific definition.

> the territory within a city contiguous to and including a highway, not comprising a business, suburban, or school district, where forty percent or more of the frontage on such highway for a distance of three hundred feet or more is occupied by dwellings or by dwellings and buildings in use for business

That is nonstarter to expect mappers to reliably apply that definition (in very rare cases where it applies)

And just burden on translators to translate that legalese is unreasonable.

See https://wiki.openstreetmap.org/wiki/Default_speed_limits for more gory details if anyone is interested. There is far more of such fun.

Asking mappers to distinguish those is just nonworkable, but tagging for "no signed speed limit here, default speed limits applies" is still possible for them and is helpful to record. And it is helpful as it points rare mappers interested in and capable of dealing with this rules to tag it somehow in more detail. And it also allows to avoid listing such roads as "speed limit missing, please survey".

Poland has much simpler distinction but distinguishing reliably between PL:urban and PL:rural is far from easy. Among complications:
- sometimes you need to survey extra kilometer to check whether urban/rural transition is signed
- PL:urban is intended to be signed also in more densely populated village areas
- sometimes it is not signed in dense built-up areas
- sometimes it is signed in empty fields or forests

giving mappers option "no signed speed limit" would limit applying of wrong PL:urban and PL:rural, while not blocking mapping by more experienced mappers.

And if you are editing tags directly then have fun remembering that in Poland "dual carriageway with 2 or more lanes in each direction" and "dual carriageway with 1 lane in each direction" get different default speed limits, in Alaska road with asphalt or concrete surface get different speed limits depending on whether they are in an unincorporated area and so on.
Escape hatch in form of being able to tag "default speed limit of some jkind applies" would help.

The same applies in many more places.

-------

So I am proposing `speed_limit_implied=yes` as new tag for that, but I am also open to better names if anyone see one.

In some areas it would exceptional and rare and in some it may be more commonly applicable. In some it may be likely and possible to fix, in some it may be unlikely to be replaced by more specific tagging.

-------

I am asking this as I am considering working on maxspeed overlay in StreetComplete that would be displaying various already tagged speed limits, highlight missing ones and allow to add missing and edit existing. 
Doing it is mammoth-sized task due to immense complexity. And I judge it to be not viable at all to implement properly without key to mark "no speed limit signs here".
By "properly" I mean here that tagging when used by people who are not experts on speed limit law can reliably improve quality of OpenStreetMap data. Obviously, completely avoiding bad edit is impossible but it should be kept to a very low rate.

See https://github.com/streetcomplete/StreetComplete/issues/4282 for issue tracking this idea.

-------

Alternative solutions:

(1) Do not implement speed limit overlay in StreetComplete if it requires such tagging, close feature request with rejection.

(2) Disable speed limit overlay by default, enable only in areas where even without using this tag it is likely that useful answer will be given by experts.
But given how much work such feature would require: outcome is probably the same as "give up on such feature"

(3) Ask local communities like USA whether they would welcome such tagging, and have speed limit overlay enabled only there and enable it by default.
It would be disabled completely in other areas.

(4) use different key, maybe even put it into maxspeed_type (`maxspeed:type=speed_limit_implied` ?). Start using [`maxspeed=implicit`](https://wiki.openstreetmap.org/wiki/Tag:maxspeed%3Dimplicit) on roads (`highway=lines`)?

(4b) use fixme - but fixme is for freeform tags, and this info in some cases would be also useful to automatically deterministic

-------

poll:
Would it be fine for StreetComplete to start using `speed_limit_implied=yes` ?
it seems fine for StreetComplete to start using `speed_limit_implied=yes`
no, StreetComplete should not start using `speed_limit_implied=yes`

disclaimer: I proposed something like that already [in January 2021 on tagging mailing list](https://lists.openstreetmap.org/pipermail/tagging/2021-January/thread.html#57929), and there was no ethusiastic support. Either it is in fact a bad idea, or I explained it poorly or people who supported it had no real option to express support. Or there was other reason for that.
I was writing it partially based on feedback from that thread.

--------

If anyone is still reading: thanks! Comments, votes, reactions are welcomed!

</details>

--------------------------

end of planned forum post

------------------------

some parts of it based on discussion so far:

- numeric speed limit in km/h (maxspeed=50) and decide on grouping
- miles per hour
- maxspeed:forward / maxspeed:backward
- symbolic (maxspeed=FR:urban)
- how to show symbolic values to user?
- numeric and symbolic on the same object - maxspeed:type=FR:urban and maxspeed=50
- mismatching numeric and symbolic on the same object - treat as missing data? ( maxspeed:type=FR:urban and maxspeed=150 )
- while we are at it, how to detect each expected value for each symbolic value? just ignore numeric if symbolic is tagged? what to display for user?
- oh, and symbolic values can change meaning - how we can detect that? rely on issue reports?
- Different speed limits per lane ( https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1570864904 )
- different speed limit per vehicle type or limited in time/whatever - entire maxspeed:conditional
- multiple tagging schemes for the same (maxspeed:type=sign source:maxspeed=DE:zone30 maxspeed:type=DE:zone30 may be used in the same case)
    - and we have bunch of other relevant taggings schemes, just a sample: source:maxspeed=DE:bicycle_road, zone:maxspeed=DE:30 , traffic_sign=DE:274.1, traffic_sign=DE:274-30, traffic_sign=DE:244.1, traffic_sign=DE:244.3
    - what about tagging maxspeed as walk? for example where traffic_sign=DE:239,1022-10 implies cycling at walking pace (i.e. maxspeed:bicycle=walk) or highway=living_street with speed limit implied.
- output likely should match whatever speed limit quest is doing, but input needs to support variety of tagging schemes
- variable speed limit display
- variable speed limit display always displaying the same speed limit
- match speed limits to our blind-friendly colour palette
- show traffic sign nodes in overlay? In more complex cases it can be handy
- support deleting/editing/adding traffic sign nodes? what about other traffic signs? Especially if tagged on one object with speed limit sign

I guess the question at which point StreetComplete should give up and announce that it is too complex to edit with it.
For example conditional per lane speed limits may be beyond point where detecting and giving up would be good enough.

see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1552141790 and https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1610257332 for UI design
see https://josm.openstreetmap.de/wiki/Styles/Maxspeed
see https://github.com/streetcomplete/StreetComplete/issues/1998#issuecomment-676787192 for some not processed info
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1546908458 for some not processed info
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1552129756 for some not processed info
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1585511029 and direct comments for some not processed info
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1969690547 for some potentially useful code and for some not processed info

https://wiki.openstreetmap.org/wiki/Default_speed_limits is an useful resource

**Comment by RubenKelevra** [comment]
> I processed large part, but not all, as I realised that without some way to mark "no explicit speed limit sign here, not sure which default rule applies" this seems not doable.

I mean we got the very unspecific established value `maxspeed=implicit` for that? 🤔

> mismatching numeric and symbolic on the same object - treat as missing data? ( maxspeed:type=FR:urban and maxspeed=150 )

I've still did not come across a good argument why we even *should* use both? I mean, what's the benefit of using a tag like `maxspeed:type=*`?

We don't tag anything else right now, which I can think off, with two tags and it adds redundancies and is really hard to parse.

I mean, say there's an implicit speedlimit in one direction and an explicit speedlimit in the other direction.

With just maxspeed=* tags we can just tag this with:

```
maxspeed:forward=FR:urban
maxspeed:backward=30
```

How would that work with maxspeed type? Add just `maxspeed:type=*` or do we use then `maxspeed:type:forward=*` as well? 

And how about restrictions in time, with one tag we would simply say:

```
maxspeed=FR:urban
maxspeed:conditional=30 @ 19:00-06:00
```

How is this supposed to work with maxspeed:type? The `maxspeed:conditional` is only applying the condition to the `maxspeed` tag, nothing else.

So logically we would be required to add another conditional tag for the type, further adding redundancies:

```
maxspeed=50
maxspeed:conditional=30 @ 19:00-06:00
maxspeed:type=FR:urban
maxspeed:type:conditional=sign @ 19:00-06:00
```

**Comment by matkoniecz** [comment]
> I mean we got the very unspecific established value maxspeed=implicit for that? 🤔

not established for use on roads themselves:

https://wiki.openstreetmap.org/wiki/Tag:maxspeed%3Dimplicit
https://overpass-turbo.eu/s/27Lo

> I've still did not come across a good argument why we even should use both? I mean, what's the benefit of using a tag like maxspeed:type=*?

I am not proposing to tag it, but still it is present in many cases.

And for why some do this - that allows data consumers to just use numeric maxspeed without maintaining or using lookup tables for which code implies which speed limit. And some data consumers are unable to consume such codes.

**Comment by matkoniecz** [comment]
I mentioned `maxspeed=implicit` in 

> (4) use different key, maybe even put it into maxspeed_type (maxspeed:type=speed_limit_implied ?). Start using [maxspeed=implicit](https://wiki.openstreetmap.org/wiki/Tag:maxspeed%3Dimplicit) on roads (highway=lines)?

(and thanks for a review!)

**Comment by RubenKelevra** [comment]
> not established for use on roads themselves:

Ah sorry, I've confused it with `maxspeed=unposted` :)

**Comment by RubenKelevra** [comment]
> And for why some do this - that allows data consumers to just use numeric maxspeed without maintaining or using lookup tables for which code implies which speed limit. And some data consumers are unable to consume such codes.

But tagging redundant information just for *some* router is considered bad practice.

"Dealing with routers that have limited support for certain common tags and use-cases is frustrating, but resorting to deceit is not the solution."

-- [Source](https://wiki.openstreetmap.org/wiki/Tagging_for_the_router)

Apart from that, data consumers always need to maintain lookup tables, which are maintained in the wiki explicitly for this purpose of being read by a machine. Otherwise, they won't be able to function in countries like Russia, Belarus, Serbia, Romania and France, which have hundreds of thousands of roads mapped with implicit tags - or otherwise they would do non-optimal routing decisions.

OsmAnd for example supports those tags for years.

**Comment by matkoniecz** [comment]
> Ah sorry, I've confused it with maxspeed=unposted :)

Does not seem to be really documented, used much (see https://wiki.openstreetmap.org/w/index.php?search=Tag%3Amaxspeed%3Dunposted&title=Special%3ASearch&profile=default&fulltext=1&searchToken=460kd2livz6indzkf53ndxhgg ) and is listed at https://wiki.openstreetmap.org/wiki/Key:maxspeed#Possible_tagging_mistakes

**Comment by RubenKelevra** [comment]
I mean it better not be used much, as it's basically like a "fixme" tag. But "there's no speedlimit posted", which is what this tag implies, is basically exactly what you're trying to convey. So it need to be surveyed for the right implicit speed limit by a more experienced mapper.

**Comment by matkoniecz** [comment]
> But tagging redundant information just for some router is considered bad practice.

oh, I am not claiming that it is a good idea

but it exists on many objects so if implementing overlay it needs to be decided what to do in such cases

**Comment by RubenKelevra** [comment]
> oh, I am not claiming that it is a good idea
> 
> but it exists on many objects so if implementing overlay it needs to be decided what to do in such cases

Like with all bad practices or outdated tags: Ask the user to confirm the current status, and then update with a check_date and just `maxspeed=*` tag, I guess? 

We did that before on some cycleway tags IIRC, which got voted to be deprecated, so SC showed a lot of "what cycleway is present there" again for me in my area.

Anyway: Thanks a lot for the work doing the discussion, that we got at least 5 standards for tagging the same information is extremely frustrating for data consumers.

**Comment by matkoniecz** [comment]
posted https://community.openstreetmap.org/t/proposing-an-new-speed-limit-implied-yes-tag-for-roads-with-default-speed-limits-where-for-some-reason-mapper-was-unable-to-be-more-specific/132700

**Comment by matkoniecz** [comment]
Also, would it be useful/possible to implement speed limit overlay that is disabled in USA and other countries with complex default speed limits?

And have it enabled only where support for them is known to be viable?

**Comment by matkoniecz** [comment]
Another option is it relax "need to be actually answerable" requirement and allow some roads to be not solvable at all. That does not seem to be a good idea to me.

**Comment by RubenKelevra** [comment]
> Another option is it relax "need to be actually answerable" requirement and allow some roads to be not solvable at all. That does not seem to be a good idea to me.

I agree, that's a bad idea.

**Comment by matkoniecz** [comment]
a bit radical and silly idea: maybe allow display of mapped speed limit signs? Maybe even allow adding new ones?

But while it would be helpful it can degenerate in horrific nightmare. Especially in cases where multiple signs are tagged in one place or on one object.

updated TODO list:

I started reviewing discussion that happened so far, to list what is required to do this behemoth.
I processed large part, but not all, as I realised that without some way to mark "no explicit speed limit sign here, not sure which default rule applies" this seems not doable.


See https://github.com/streetcomplete/StreetComplete/issues/1998#issuecomment-676787192 and https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1969690547 for why two previous attempts to support max speed in such way was abandoned.

some parts of it based on discussion so far (likely some should be dropped from final version):

- in order to not simply overwrite previous taggings with for example `maxspeed:type`, the app would need to understand and treat correctly all of `source:maxspeed`, `maxspeed:type`, `zone:maxspeed` and depending on which tag is already set on the road, specifically update this value with the surveyed value. Also their subvariants for traffic direction, conditional tagging, vehicle type subtags, per-lane subtags and so on.
- output likely should match whatever speed limit quest is doing, but input needs to support variety of tagging schemes 
- we need way to disable/enable overlays per country (like quests can be)
- support display of numeric speed limit in km/h (maxspeed=50)
- decide on grouping them in displayed colours (our blind-friendly colour palette)
- the same for miles per hour speed limit
- decide what counts as suspicious value for max speed 
- support speed limit zones
- maybe try drawing speed limit zones blobs on the map
- maxspeed:forward / maxspeed:backward
- symbolic (`maxspeed=FR:urban` `maxspeed:source=FR:urban` `maxspeed:type=FR:urban` and other values, see https://taginfo.openstreetmap.org/keys/maxspeed)
- so speed limit and its type (sign, default speed limit) - may be tagged directly in maxspeed (`maxspeed=RO:urban`_ or in `source:maxspeed`  ( https://taginfo.openstreetmap.org/keys/source%3Amaxspeed#values ) or in `maxspeed:type` ( https://taginfo.openstreetmap.org/keys/maxspeed:type#values )
- and both read and write support is needed for and all this urban/rural, and other fixed speed limits based on another sign (living streets, bicycle roads?, pedestrian zone?, school zone?, silver zone?, shared zone?, ... more? - see default speed limits, depending on the country, there may be many of these). Of course, need (simplified) pictures of such signs too for interface
- how to show symbolic values to user? As symbols? With numeric values implied by that symbolic values? Both?
- numeric and symbolic on the same object - maxspeed:type=FR:urban and maxspeed=50
- mismatching numeric and symbolic on the same object - treat as missing data? ( maxspeed:type=FR:urban and maxspeed=150 ) just ignore numeric if symbolic is tagged? what to display for user?
- while we are at it, how to detect each expected value for each symbolic value? 
- expected speed limit indicated by `FR:rural` depends on exact situation, and requites taking other tags into account. For example it may change if road is dual carriageway - and while `oneway=yes` is nice heuristic it is not foolproof - see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1585511029 (I guess if `maxspeed` is tagged then we can check is it possible for it to match... And raise alarm only if it definitely not matches. What to show when there are multiple options?)
- integrating with `osm-legal-default-speeds`
- review/update `osm-legal-default-speeds` for missed legislation and legislation changes
- deal with https://community.openstreetmap.org/t/inconsistent-definition-of-vehicle-categories-between-default-speed-limits-and-their-respective-wiki-pages/133330 mismatch
- symbolic values can change meaning when legislation changes - how we can detect that? rely on issue reports?
- multiple tagging schemes for the same (maxspeed:type=sign source:maxspeed=DE:zone30 maxspeed:type=DE:zone30 may be used in the same case)
- Different speed limits per lane ( https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1570864904 )
- different speed limit per vehicle type (in order of lower usage: maxspeed:hgv maxspeed:bus maxspeed:trailer maxspeed:hgv:conditional maxspeed:motorcar maxspeed:psv:conditional maxspeed:goods maxspeed:coach maxspeed:motorcycle )
- we need icons for this vehicle types
- speed limits for specific hours - subset of maxspeed:conditional - see https://taginfo.openstreetmap.org/keys/maxspeed%3Aconditional#values and https://wiki.openstreetmap.org/wiki/Conditional_restrictions#Speed_limit
- so parser of subset of maxspeed:conditional is needed
- and GUI to display it
- and GUI to allow edits on it
- more conditional speed limits (no support?)
- and we have bunch of other relevant taggings schemes, just a sample: source:maxspeed=DE:bicycle_road, zone:maxspeed=DE:30 , traffic_sign=DE:274.1, traffic_sign=DE:274-30, traffic_sign=DE:244.1, traffic_sign=DE:244.3
- what about tagging maxspeed as walk? for example where traffic_sign=DE:239,1022-10 implies cycling at walking pace (i.e. maxspeed:bicycle=walk) or highway=living_street with speed limit implied, which in some (not all!) area will be `maxspeed=walk`.
- oh, also "unlimited speed limit" on German autobahn needs support
- variable speed limit display ( maxspeed:variable )
- reasons for variable speed limit display ( maxspeed:variable )
- which of `maxspeed:variable:forward` / `maxspeed:variable:backward` (only in one direction), per lane tagging, specific for given vehicle type, only on specific hours, other conditional speed limit need to be implemented for `maxspeed:variable` ?
- variable speed limit display always displaying the same speed limit (how to mark it? maybe just showing `note` is enough - see https://community.openstreetmap.org/t/how-to-map-that-led-displayed-speed-limit-is-not-really-variable/133362 )
- show traffic sign nodes in overlay? In more complex cases it can be handy
- support deleting/editing/adding traffic sign nodes? what about other traffic signs? Especially if tagged on one object with speed limit sign
- `maxspseed:advisory`
- `maxspseed:advisory:forward` / `maxspseed:advisory:backward`  (but it is less than 2k cases globally), see https://taginfo.openstreetmap.org/search?q=maxspeed%3Aadvisory
- `maxspeed:advisory:lanes` ? less than 1000 cases globally?
- `maxspeed:advised` duplicate probably can be ignored? But I will try to eliminate it, there is no reason whatsoever for it to exist
- can something have regular speed limit and advisory speed limit? Probably yes, because world is cursed. How to handle it?
- BTW, advisory speed limit seems to use `source:maxspeed:advisory` to record source more often than `source:maxspeed` does
- how to handle say `zone:maxspeed=DE:urban` that is to be marked as having specific traffic sign limiting speed to 30 km/h? Set `maxspeed:type=sign` and remove `zone:maxspeed` rather than setting `zone:maxspeed=sign` - right? Initial replies to https://community.openstreetmap.org/t/what-is-the-point-of-zone-traffic/133391 disagree
- how to deal with freeform `source:maxspeed` values? Including ones actually used to record source of edit such as `source:maxspeed=https://portal-mainroads.opendata.arcgis.com/datasets` (see [taginfo](https://taginfo.openstreetmap.org/tags/?key=source%3Amaxspeed&value=https%3A%2F%2Fportal-mainroads.opendata.arcgis.com%2Fdatasets))
- start thread about `source:maxspeed` used to store actual maxspeed source, not its type
- [source:maxspeed](https://taginfo.openstreetmap.org/keys/source%3Amaxspeed)=practical is apparently a thing (but looking at https://taginfo.openstreetmap.org/tags/source%3Amaxspeed=practical it seems some bad import/mass edit, looking at https://overpass-turbo.eu/s/28SN it seems to be a local aberration and in https://www.openstreetmap.org/changeset/34906196 I am attempting to contact whoever added it)
- on user selecting living street either existing `maxspeed` should be removed (including `maxspeed=walk` and `maxspeed=20` which may be valid, depending on location) or app need to know which speed limit applies to living street in each place where living street exist (and maintain it, following legislation changes)
- BTW, `living_street=yes` exists
- `source:maxspeed=RU:living_street` appears to apply not only on living streets - see https://github.com/streetcomplete/StreetComplete/issues/1998#issuecomment-676787192
- how to deal with speed limits on railways, cycleways and waterways? - ignore, probably. But what about bicycle roads and bicycle boulevards and pedestrian streets?
- how to deal with `maxspeed:forward` `maxspeed:backward` if tagged on oneways (note though `oneway:bicycle`)
- see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-3133552716 for rough plan
- reserve time on communicating with OSM community (this would explicitly exclude time on changing OSM tagging or bot edits)
- `zone:traffic` is not actually holding speed limits, but if `??:urban` is answered on `zone:traffic=??:rural` then `zone:traffic`  should be updated (the same for other cases where user gave answer definitely contradicting claimed `zone:traffic` )  (see https://community.openstreetmap.org/t/what-is-the-point-of-zone-traffic/133391 )
- is `zone:maxspeed` behaving in the same way? See https://community.openstreetmap.org/t/is-zone-maxspeed-de-30-maxspeed-10-maxspeed-type-sign-potentially-valid/133430
- `maxspeed:practical` and its cousins will be ignored altogether - in no case it needs to be changed, also it is subjective, unverifiable and problematic tag that IMHO should not exist
- trawl taginfo for maxspeed keys to detect what was missed here
- note that some areas may have extra bonuses like cryptic zone signs (in Poland if regular non-zone speed limit sign is attached in some specific way to settlement sign it becomes speed limit zone sign)
- which maxspeed tags should be removed when speed limit changes?
- remove speed limit quest, includes required migration of internal database and statistics, test how it works for people who have not yet uploaded speed limit quests
- maybe reuse kotlin parser and have it parse all speed limits tagged in OSM and report self-contradictions? and clearly bad tagging? and publish list?
- test speed limit overlay on people (UX test) - is it clear enough to have it enabled by default? Or does it require hiding like MTB overlay. Hopefully it being an overlay that needs to be enabled is enough of gating
- reserve time for code review


I guess the question at which point StreetComplete should give up and announce that it is too complex to edit with it.
For example conditional per lane speed limits may be beyond point where detecting and giving up would be good enough.

integrate https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-3133441038
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1552141790 and https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1610257332 for UI design
see https://josm.openstreetmap.de/wiki/Styles/Maxspeed
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1585511029 and direct comments for some not processed info
see https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-1969690547 for some potentially useful code - they got stuck on conditional speed limits

https://wiki.openstreetmap.org/wiki/Default_speed_limits is an useful resource

**Comment by matkoniecz** [comment]
@westnordost 
would it be fine to refuse edit some relatively rare cases? Or do we need to support lane-specific conditional expressions? Or any freeform conditional expressions?

Hopefully not, because that would triple effort to get thousands additional objects supported.

----------------------------------------------------------------------------

What about when different tagging schemes disagree? 

`maxspeed=100 source:maxsped=PL:rural type:maxspeed=PL:urban` 

would it be fine to either refuse editing there OR to pick up random tagging of placed? Because having comprehensible interface displaying all conflicting info seems a bit too much.

**Comment by RubenKelevra** [comment]
> Or do we need to support lane-specific conditional expressions? Or any freeform conditional expressions?

I don't think so, but maxspeed based on directions of travel would be really nice, as it cuts the work down in half for mapping a single way. Meaning you can map one side, and don't have to survey everything in the other direction, too just to map it. 

Here in my area we got many areas where we got different speedlimits in sections for different directions of travel, due to corners being to tight in one direction etc.

**Comment by matkoniecz** [comment]
looking at https://github.com/streetcomplete/StreetComplete/issues/1998#issuecomment-676787192

1. seems annoying but doable
2. seems doable if user is expected to mark whether given sign is a zone sign or non-zone sign (but that can be nightmare with some signs being zone one, despite not using zone design - for example in Poland speed limit sign attached in specific way to settlement name sign becomes zone sign...)
3. should be solved by kicking out USA and similarly problematic areas from speed limit overlay - or is similar weirdness able to appear anywhere?
4. seems doable?
5. irritating but also seems doable (BTW, such info needs to be collected not only for speed limit implied by `living_street` but probably also for `RO:rural` `PL:urban` and so on....)
6. small nit but also doable

**Comment by RubenKelevra** [comment]
> What about when different tagging schemes disagree?
> 
> `maxspeed=100 source:maxsped=PL:rural type:maxspeed=PL:urban`

Best solution IMHO is to show a resurvey to the user and delete everything except the new tags in this case.

**Comment by matkoniecz** [comment]
> Best solution IMHO is to show a resurvey to the user and delete everything except the new tags in this case.

but what to show user when they will tap on road? nothing? random one? Is it fine?

> I don't think so, but maxspeed based on directions of travel would be really nice,

definitely planned, see `maxspeed:forward / maxspeed:backward` - this is one of less complex parts

**Comment by arrival-spring** [comment]
You could also throw `maxspeed:practical` into the mix.

___

I wonder if the best way is just to make an overlay for a single country (at a time) that covers 95-99% of cases, leaving the rest uneditable, rather than trying to solve all cases all at once.

e.g. in the UK for this you would need
`maxspeed:type` (/`source:maxspeed`) = 
* GB:nsl_single
* GB:nsl_dual
* GB:nsl_restricted
* GB-WLS:nsl_restricted
* GB:motorway
* GB:zone20
* GB:zone40
* sign

And then maxspeed = 10 mph to 70 mph in 10 mph increments.

This covers almost everything. I would go with a colouring for the speeds, but if any type was set then colour based on that instead.

If type and speed is set then show both to the user using icons/minimal text.

If nothing is set then do what the quest does.

Nice to haves would be `maxspeed:variable` and `maxspeed:advisory` (which is found in addition to the set speed limit).

Different forward and backward speed limits are very rarely found in the UK.

Just saying this to explain that a single country could be fairly simple, it's when you try to deal with all cases that it gets too complicated.

**Comment by RubenKelevra** [comment]
@arrival-spring can you explain how two mappers visiting a road can come to the same conclusion regarding `maxspeed:practical`?


> e.g. in the UK for this you would need
> `maxspeed:type` (/`source:maxspeed`) =

Why would you need an additional tags like `maxspeed:type` (/`source:maxspeed`) to map this? You can tag those values directly as `maxspeed=*`.

I don't understand the need for redundancy in this case, apart from the fact that if the law changes, like recently I Spain all roads need to be changed, by updating the `maxspeed=*` number because it's now in disagreement with the `maxspeed:type=*` / `source:maxspeed=*`.

**Comment by matkoniecz** [comment]
> You could also throw maxspeed:practical into the mix.

not adding this one to pile - as it is subjective, unsigned and not really verifiable and orthogonal to signed speed limit I am happy that this one can be ignored rather than making nightmare worse

> I wonder if the best way is just to make an overlay for a single country (at a time) that covers 95-99% of cases, leaving the rest uneditable, rather than trying to solve all cases all at once.

my plan would be to start from something even simpler - end to end tool that "just" parses basic tagging and shows raw tags and is able to save edits and try using it, add very basic support for PL:rural PL:urban and speed zones and their display

if such minimum viable tool will turn out to not be useful in practice, then drop entire project

if it will turn out to be working well, then collect various test cases for parser, build test suite, add fuller parser. Parser would be able to detect that some roads are beyond interface to support and show raw tag list in such case (or block editing completely), then add feature one by one such as lane-specific tagging and other complex stuff

the question is how large coverage of countries is expected to be before opening PR - would `maxspeed` overlay supporting solely say Poland and Germany would be fine for PR and merge with next countries and their features added in subsequent PRs? Or is almost-global coverage expected for initial PR? Maybe even Poland-only would be fine for start?

(oh and before implementing anything: finish already open PRs first)

**Comment by arrival-spring** [comment]
Oh I didn't mean that it should be mapped in the overlay, but is something to think about since it's been used more than maxspeed:lanes.

When I was thinking about making an overlay my thought was to remove it if a new speed limit was entered that was lower than an existing practical limit, otherwise to leave it alone.

**Comment by matkoniecz** [comment]
> When I was thinking about making an overlay my thought was to remove it if a new speed limit was entered that was lower than an existing practical limit, otherwise to leave it alone.

it seems that even then people may mark it for cases where everyone goes over speed limit

and my position would be to avoid touching it at all, to reduce complexity and because it seems subjective, unverifiable and highly problematic tag that in my opinion should not exist
so I prefer to ignore its existence rather than support it

I will mention it in the list

**Comment by matkoniecz** [comment]
`zone:traffic` is not actually holding speed limits, but if `??:urban` is answered on `zone:traffic=??:rural` then `zone:traffic` should be updated (the same for other cases where user gave answer definitely contradicting claimed `zone:traffic` )

at least according to https://community.openstreetmap.org/t/what-is-the-point-of-zone-traffic/133391

I updated https://github.com/streetcomplete/StreetComplete/issues/4282#issuecomment-3133408120 accordingly (among many other updates)

**Comment by matkoniecz** [comment]
> speed limit value in a square in case of a zone:mawspeed=* tag

Turns out that `zone:maxspeed` is maybe not actually for max speed, see https://community.openstreetmap.org/t/is-zone-maxspeed-de-30-maxspeed-20-maxspeed-type-sign-potentially-valid/133430
